ukrainian_word_stress/stressify_.py

```diff
@@ -0,0 +1,210 @@
+            - `OnAmbiguity.Skip` (default): do not place stress
+            - `OnAmbiguity.First`: place a stress of the first match with a
+                high chance of being incorrect.
+            - `OnAmbiguity.All`: return all possible options at once.
+                This will look as multiple stress symbols in one word
+                (за´мо´к)
+
+    Example:
+        >>> f = Stressifier()
+        >>> f("Привіт, як справи?")
+        'Приві´т, як спра´ви?'
+    """
+
+
+
+    def __init__(self,
+                 stress_symbol=StressSymbol.AcuteAccent,
+                 on_ambiguity=OnAmbiguity.Skip):
+
+        dict_path = pkg_resources.resource_filename('ukrainian_word_stress', 'data/stress.v2.trie')
+        self.dict = marisa_trie.BytesTrie()
+        self.dict.load(dict_path)
+        self.nlp = stanza.Pipeline(
+            'uk',
+            processors='tokenize,pos,mwt',
+            download_method=stanza.pipeline.core.DownloadMethod.REUSE_RESOURCES,
+            logging_level=logging.getLevelName(log.getEffectiveLevel())
+        )
+        self.stress_symbol = stress_symbol
+        self.on_ambiguity = on_ambiguity
+
+    def __call__(self, text):
+        parsed = self.nlp(text)
+        result = MutableText(text)
+        log.debug("Parsed text: %s", parsed)
+        for token in parsed.iter_tokens():
+            accents = find_accent_positions(self.dict, token.to_dict()[0], self.on_ambiguity)
+            accented_token = self.apply_accent_positions(token.text, accents)
+            if accented_token != token:
+                result.replace(token.start_char, token.end_char, accented_token)
+
+        return result.get_edited_text()
+
+    def apply_accent_positions(self, s, positions):
+        for position in sorted(positions, reverse=True):
+            s = s[:position] + self.stress_symbol + s[position:]
+        return s
+
+
+def stressify(text: str) -> str:
+    """Add word stress to texts in Ukrainian.
+
+    This is a functional interface to `Stressifier`.
+
+    Example::
+        >>> stressify("Привіт, як справи?")
+        'Приві´т, як спра´ви?'
+
+    """
+
+    if not hasattr(stressify, "f"):
+        stressify.f = Stressifier()
+    return stressify.f(text)
+
+
+def find_accent_positions(trie, parse, on_ambiguity=OnAmbiguity.Skip) -> List[int]:
+    """Return best accent guess for the given token parsed tags.
+
+    Returns:
+        A list of accent positions. The size of the list can be:
+        0 for tokens that are not in the dictionary.
+        1 for most of in-dictionary words.
+        2 and more - for compound words and for words that have
+          multiple valid accents.
+    """
+
+    base = parse['text']
+    for word in (base, base.lower(), base.title()):
+        if word in trie:
+            values = trie[word]
+            break
+    else:
+        # non-dictionary word
+        log.debug("%s is not in the dictionary", base)
```

Comment:
you can still accent it, if it has just one vowel, no?

Reply:
Right. But since we don't show accent for single-syllable words, the end result will be the same.